xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -830,26 +830,29 @@ private static CdsUpdate processCluster(Cluster cluster, Set<String> retainedEds
         break;
       case CLUSTERDISCOVERYTYPE_NOT_SET:
       default:
-        throw new ResourceInvalidException("Unspecified cluster discovery type");
+        throw new ResourceInvalidException(
+            "Cluster " + cluster.getName() + ": unspecified cluster discovery type");
     }
     if (structOrError.getErrorDetail() != null) {
       throw new ResourceInvalidException(structOrError.getErrorDetail());
     }
-
     CdsUpdate.Builder updateBuilder = structOrError.getStruct();
 
     if (cluster.getLbPolicy() == LbPolicy.RING_HASH) {
       RingHashLbConfig lbConfig = cluster.getRingHashLbConfig();
-      if (lbConfig.getHashFunction() != RingHashLbConfig.HashFunction.XX_HASH) {
+      long minRingSize = lbConfig.getMinimumRingSize().getValue();
+      long maxRingSize = lbConfig.getMaximumRingSize().getValue();
+      if (lbConfig.getHashFunction() != RingHashLbConfig.HashFunction.XX_HASH
+          || minRingSize > maxRingSize) {
```

Comment:
Your change in `CdsLoadBalancer2.java` seems to allow both `minRingSize` and `maxRingSize` to be zero, but here it is not allowed?